### PR TITLE
=doc create new `@signature` directive to support directive documentation

### PR DESCRIPTION
**Pull request by jrudolph**
This is basically a port from the old `includecode2.py` script customization we built for spray.

I guess that's as good as it gets without spending too much time on it if #11 is the way to go eventually.

WDYT, @jonas?

### CODE REVIEW

File: project/ParadoxSignatureDirective.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
+package akka
+
+import sbt._
+import java.io.{File, FileNotFoundException}
+
+import com.lightbend.paradox._
+import markdown.Writer
```
Comment (jonas) [inline project/ParadoxSignatureDirective.scala]: Looks like this is imported by `com.lightbend.paradox.markdown.{LeafBlockDirective, _}` below

File: project/ParadoxSignatureDirective.scala [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        //println(s"Looking for signature regex '$Signature'")
+        val text =
+          Source.fromFile(file).getLines.collect {
+            case line@Signature(full, l, _*) if labels contains l =>
+              //println(s"Found label '$l' with sig '$full' in line $line")
+              full
+          }.mkString("\n")
+
+        val lang = Option(node.attributes.value("type")).getOrElse(Snippet.language(file))
+        new VerbatimNode(text, lang).accept(visitor)
```
Comment (jonas) [inline project/ParadoxSignatureDirective.scala]: This should fail if text is empty

**Comment by jonas** [comment]
FWIW, LGTM

File: project/ParadoxSignatureDirective.scala (hunk 1 above)
Comment (jrudolph) [inline project/ParadoxSignatureDirective.scala]: Thanks, will do.

File: project/ParadoxSignatureDirective.scala (hunk 2 above)
Comment (jrudolph) [inline project/ParadoxSignatureDirective.scala]: Yes, good point.

Comment (jrudolph) [inline project/ParadoxSignatureDirective.scala]: Ha, this directly catched a problem when referencing a type alias in `authenticateBasic.md`.

**Comment by jrudolph** [comment]
Thanks for the comments, @jonas. I've now added a warning when the directive doesn't match at all (otherwise, if we'd threw an exception, we'd need paradox support to make it report multiple failures. Also we inject error text directly into the document.

We can now only include `type` definitions in which case the complete definition is included and not just the signature. Maybe this extra handling could be replaced by more generic parameter to the directive whether to include the complete line but as we still essentially want to get rid of the directive pages here, I guess it doesn't matter too much.

**Comment by jonas** [comment]
👍 Paradox' embrace of "let it crash" on first failure was not fun during the doc migration.

File: project/ParadoxSignatureDirective.scala [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            case line@Signature(signature, kind, l, definition) if labels contains l.toLowerCase() =>
+              //println(s"Found label '$l' with sig '$full' in line $line")
+              if (kind == "type") signature + definition
+              else signature
+          }.mkString("\n")
+
+        if (text.trim.isEmpty) {
+          logWarn(
+            s"Did not find any signatures with one of those names [${labels.mkString(", ")}] in ${node.source} " +
+            s"(was referenced from [${page.path}])")
```
Comment (ktoso) [inline project/ParadoxSignatureDirective.scala]: Ok, though in final versions I'd prefer to fail (accumulate many errors and fail) but paradox can't do that right now.

**Comment by ktoso** [comment]
LGTM, nice - thanks @jrudolph !

**Pull request merged**